vpr/src/base/read_place.cpp

```diff
@@ -99,192 +150,145 @@ void read_place(const char* net_file,
-            VPR_THROW(VPR_ERROR_PLACE_F, constraints_file, vtr::get_file_line_number_of_last_opened_file(),
-                      "Block %s subtile number (%d) is out of range.\n", bname, k);
-        }
-        place_ctx.grid_blocks[i][j].blocks[k] = bnum;
-        place_ctx.grid_blocks[i][j].usage++;
+            //Set the location
+            place_ctx.block_locs[blk_id].loc.x = block_x;
+            place_ctx.block_locs[blk_id].loc.y = block_y;
+            place_ctx.block_locs[blk_id].loc.sub_tile = sub_tile_index;
 
-        constrained_blocks.insert(bnum);
+            //Check if block is at an illegal location
 
-        ptr = vtr::fgets(buf, vtr::bufsize, fp);
-    }
+            auto physical_tile = device_ctx.grid[block_x][block_y].type;
+            auto logical_block = cluster_ctx.clb_nlist.block_type(blk_id);
+
+            if (sub_tile_index >= physical_tile->capacity || sub_tile_index < 0) {
+                VPR_THROW(VPR_ERROR_PLACE, "Block %s subtile number (%d) is out of range. \n", c_block_name, sub_tile_index);
+            }
+
+            if (!is_sub_tile_compatible(physical_tile, logical_block, place_ctx.block_locs[blk_id].loc.sub_tile)) {
+                VPR_THROW(VPR_ERROR_PLACE, "Attempt to place block %s with ID %d at illegal location (%d, %d). \n", c_block_name, blk_id, block_x, block_y);
+            }
+
+            //need to lock down blocks  and mark grid block usage if it is a constraints file
+            if (!is_place_file) {
+                place_ctx.block_locs[blk_id].is_fixed = true;
+                place_ctx.grid_blocks[block_x][block_y].blocks[sub_tile_index] = blk_id;
+                place_ctx.grid_blocks[block_x][block_y].usage++;
+            }
+
+            //mark the block as seen
+            seen_blocks[blk_id] = 1;
 
-    for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
-        auto result = constrained_blocks.find(blk_id);
-        if (result == constrained_blocks.end()) {
-            continue;
+        } else {
+            //Unrecognized
+            vpr_throw(VPR_ERROR_PLACE_F, place_file, lineno,
+                      "Invalid line '%s' in file",
+                      line.c_str());
         }
+    }
 
-        if (place_ctx.block_locs[blk_id].loc.x == OPEN) {
-            VPR_THROW(VPR_ERROR_PLACE_F, constraints_file, 0,
-                      "Block %s location was not specified in the constraints file.\n", cluster_ctx.clb_nlist.block_name(blk_id).c_str());
+    //For place files, check that all blocks have been read
+    if (is_place_file) {
+        for (auto block_id : cluster_ctx.clb_nlist.blocks()) {
+            if (seen_blocks[block_id] == 0) {
+                VPR_THROW(VPR_ERROR_PLACE, "Block %d has not been read from the place file. \n", block_id);
+            }
         }
     }
 
-    fclose(fp);
-    free_hash_table(hash_table);
-    VTR_LOG("Successfully read %s.\n", constraints_file);
-    VTR_LOG("\n");
+    if (is_place_file) {
```

Comment:
add comment to explain why

Reply:
for entire place file, want to make hash so we can write into routing file for later error checking, dont need for constraint